ckanext/ontario_theme/anti_csrf3.py

```diff
@@ -0,0 +1,121 @@
+import anti_csrf
+import webob
+from webob.exc import HTTPForbidden
+
+from flask import Flask, make_response, render_template, request
+from ckan.common import config, is_flask_request
+import ckan.lib.base as base
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+CSRF_ERR = 'CSRF authentication failed. Token missing or invalid.'
+
+domain = config.get('ckan.ontario_theme.csrf_domain', '')
+
+
+
+def after_request_function(response):
+    #request = Request(environ)
+    #self.session = environ['beaker.session']
+    #self.session.save()
+    #if is_valid():
+    resp = response
+    #else:
+    #    return response
+    if 'text/html' in resp.headers.get('Content-type', ''):
+        token = anti_csrf.get_response_token(request, resp)
+        new_response = anti_csrf.apply_token(resp.get_data(as_text=True), token)
+        resp.set_data(new_response)
+        return resp
+
+    else:
+        return response
+
+def is_valid():#
+    return is_safe() or unsafe_request_is_valid()
+
+def unsafe_request_is_valid():
+#        return is_secure() and good_referer() and \
+#               good_origin() and check_token()
+    return check_token()
+
+##############################################################################
+
+
+def is_secure():
+    # allow requests which have the x-forwarded-proto of https (inserted by nginx)
+    if request.headers.get('X-Forwarded-Proto') == 'https':
+        return True
+    return request.scheme == 'https'
+
+def is_safe():
+    "Check if the request is 'safe', if the request is safe it will not be checked for csrf"
+    # api requests are exempt from csrf checks
+    if request.path.startswith("/api"):
+        return True
+
+    # get/head/options/trace are exempt from csrf checks
+    return request.method in ('GET', 'HEAD', 'OPTIONS', 'TRACE')
```

Comment:
TRACE should probably be entirely blocked, not given special treatment. It actually enables some types of CSRF filter bypass, by exposing the token cookie.